Reduce transaction scope and heartbeat frequency

steventlamb opened the pull request:
connects #2500 on github

These two features together lead to a large increase in task runtime in
cases where multiple tasks were running in parallel. This is because
each task would hold a transaction with edits to one import_event record
for the entire life of the task. The result is that each task would have
to wait until a sibling task finishes its transaction. We were seeing
queries in excess of ten seconds as a result. By scoping the transaction to just the
length of the row edits, queries from other tasks need only block as long as it takes a
task to edit each row.

Additionally, the heartbeat was being run once per row (85 times per
task) when once per task is perfectly adequate.

It is suspected, but not confirmed, that the heartbeat is the particular
aspect of the row edit that was causing the lock on import_event that
was blocking other tasks. It is possible that performing either one of
these two changes would have fixed this issue. However, they are both
improvements so there is little benefit to further investigation.

To test:
1) run an import job with rate limiting at its current setting of
1/m. Note the time per job.
2) change rate limiting value to something like 15/m.
3) run an import job. If this fix works, the tasks run more frequently
but with the same average runtime.

RickMohr commented:
Changes look good. I'd like to understand how you measured job time before I test.

steventlamb commented:
I measured job time by:
1) running `scripts/debugcelery.sh`, which produces log lines with job times. This is the main thing to watch when analyzing job times (as opposed to system resources, for which I used Vector)
2) used `manage.py random_csv` to create a dataset of about 10k trees
3) Adjusted my vagrantfile to add more CPU cores to my app VM. Two should be fine to replicate the transaction locking.
3) Adjusted the values in `default_settings.py` for rate limiting. Depending on your system resources, you should find some key values where the situation changes. For example, at the current setting of 1/m, jobs finish fast enough and with enough sparsity that two workers don't ever block. At a setting of 100/m (effectively no rate limiting), you should see job times increase (until you checkout this branch). When I was tuning I noticed interesting things happen across the 5/m-15/m range, but that shouldn't be necessary for this test.

Please let me know if you need more info.

RickMohr commented:
For the record, we observed that with 1 cpu, `debugcelery.sh` creates two celery worker processes, so changing the number of cpus shouldn't be necessary.

RickMohr commented:
+1 At `15/m` I saw 9-10 sec/task without these changes and 6-7 sec/task with these changes.

I did need to use 2 cpus -- with 1 cpu the same process got all the tasks.
